### PR TITLE
docs: instructions for using the trainer controller framework

### PR DESCRIPTION
**Pull request by HarikrishnanBalagopal**
### Description of the change

Usage instructions for the trainer controller framework.

### Related issue number

https://github.com/foundation-model-stack/fms-hf-tuning/pull/45

### How to verify the PR

NA

### Was the PR tested

- [x] I have added >=1 unit test(s) for every new method I have added.
- [x] I have ensured all unit tests pass

### CODE REVIEW

File: examples/trainer_controller/README.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  - name: <controller-name>
+    triggers:
+      - <event-1>
+      ...
+    rule: <rule-string>
+    operations:
+      - <operation-action-1>
+      ...
+```
+The `controller-metrics` and `operations` are optional. We provide a set of built-in `controller-metrics` and `operations` which could be referred to without actually defining them as. For example, the below configuration defines a `controller-metric` called `loss` which refers to a built-in `Loss` controller-metric class with custom arguments (in this case, no arguments. If arguments are required, then they could be listed under a `arguments` section as shown above), but does not define any `operations`. It only refers to a built-in operation.
```
Comment (Ssukriti) [inline examples/trainer_controller/README.md:46]: >We provide a set of built-in `controller-metrics` and `operations` which could be referred to without actually defining them as.

where is this list?

File: README.md [hunk 2]
```diff
@@ -516,6 +536,10 @@ python main.py \
 
 The above runs several tasks with `hendrycksTest-*` being MMLU.
 
+## Trainer Controller Framework
+
+For details about the trainer controller framework see [examples/trainer_controller/README.md](examples/trainer_controller/README.md)
```
Comment (Ssukriti) [inline README.md:541]: can you move few lines explanation here `A framework for controlling the trainer loop using user-defined rules and metrics.` . This can help with optimal stopping of training loop , where user could specify rules like stop training if loss below a threshold

File: examples/trainer_controller/README.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+controller-metrics:
+  - name: loss
+    class: Loss
+controllers:
+  - name: loss-controller
+    triggers:
+      - on_log
+    rule: "loss < 1.0"
+    operations:
+      - hfcontrols.should_training_stop
```
Comment (Ssukriti) [inline examples/trainer_controller/README.md:57]: where is the list of operations that can be specified?

File: README.md (hunk 2 above)
Comment (Ssukriti) [reply README.md:541]: why will anyone open details when they dont even know what is Trainer Controller? :)

File: examples/trainer_controller/README.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+- **Events**: Events supply the state and arguments required for the metric handlers to perform metric computation at the events they are registered for. The framework callback lists out all event handlers with prefix `"on_"` and loads then as event handlers. Every metric declares one or more events from this list of valid handlers. These computed metric variables are stored in a global state of the trainer controller and independently picked up the operations which could potentially be triggered on an entirely different set of events. This decouples the control loop for metrics and operations. I.e. the metric could be computed on event A, while operation could be triggered on event B. The controller rules which use the metric variables from the trainer controller state are evaluated and based on the outcomes specified actions are performed.
+![High-Level Design Diagram: Trainer Controller Framework](../../architecture_records/imgs/001-arch.png)
+
+## Usage and customization
+We added a new trainer callback (see [here](https://huggingface.co/docs/transformers/v4.37.2/en/main_classes/callback)) which accepts a `training control definition` YAML file containing:
```
Comment (Ssukriti) [inline examples/trainer_controller/README.md:15]: how to pass this YAML file while training? can you list parameter name and show full example on how to use it with fms-hf-tuning for any one tuning type ? that parameter name can be moved to README.md and this section can be about describing YAML in more detail like you did. Moving comment above

File: README.md (hunk 2 above)
Comment (Ssukriti) [reply README.md:541]: here please add a example of how to pass the YAML to training loop . 
Example:
Sample YAML can look like this
```
controller-metrics:
  - name: loss
    class: Loss
controllers:
  - name: loss-controller
    triggers:
      - on_log
    rule: "loss < 1.0"
    operations:
      - hfcontrols.should_training_stop
```
To use it with training , command 
```
`accelerate launch \
--num_processes=4
--config_file fixtures/accelerate_fsdp_defaults.yaml \
tuning/sft_trainer.py  \
--model_name_or_path $MODEL_PATH  \
--training_data_path $TRAIN_DATA_PATH  \
--output_dir $OUTPUT_PATH  \
--num_train_epochs 5  \
--per_device_train_batch_size 4  \
--learning_rate 1e-5  \
--response_template "\n### Label:"  \
--dataset_text_field "output" \
--peft_method "None"
---trainer_controller_file = sample_yaml
```

then say for details in structure of YAML and what kind of metrics you can pass , refer to [section ](examples/trainer_controller/README.md)

File: examples/trainer_controller/README.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  - name: loss
+    class: Loss
+controllers:
+  - name: loss-controller
+    triggers:
+      - on_log
+    rule: "loss < 1.0"
+    operations:
+      - hfcontrols.should_training_stop
+```
```
Comment (Ssukriti) [inline examples/trainer_controller/README.md:58]: do you want to add another file to this folder with example YAMLs containing recommended metrics/ metrics that have been tested for use cases we know of so far?

File: examples/trainer_controller/README.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+      ...
+    rule: <rule-string>
+    operations:
+      - <operation-action-1>
+      ...
+```
+The `controller-metrics` and `operations` are optional. We provide a set of built-in `controller-metrics` and `operations` which could be referred to without actually defining them as. For example, the below configuration defines a `controller-metric` called `loss` which refers to a built-in `Loss` controller-metric class with custom arguments (in this case, no arguments. If arguments are required, then they could be listed under a `arguments` section as shown above), but does not define any `operations`. It only refers to a built-in operation.
+```yaml
+controller-metrics:
+  - name: loss
```
Comment (Ssukriti) [inline examples/trainer_controller/README.md:49]: how can I know if it is validation loss or train loss?

**Comment by seshapad** [comment]
@Ssukriti thank you very much for the comments. We will revise this document based on your comments. Also, we have made the PR draft to have it reviewed within the team.

File: README.md [hunk 7]
```diff
@@ -1,5 +1,25 @@
 # FMS HF Tuning
 
+- [Installation](#installation)
```
Comment (seshapad) [inline README.md:3]: @HarikrishnanBalagopal why is this section having so many lines of change. Only addition is line 20, isnt it?

File: README.md [hunk 8]
```diff
@@ -516,6 +536,19 @@ python main.py \
 
 The above runs several tasks with `hendrycksTest-*` being MMLU.
 
+## Trainer Controller Framework
+
+Trainer controller is a framework for controlling the trainer loop using user-defined rules and metrics.
+
+The motivation for this framework:
```
Comment (seshapad) [inline README.md:543]: Please replace line 545 through 551, with below passage:
```

This frameworks helps user define rules to capture scenarios like criteria for stopping an ongoing training (E.g validation loss reaching a certain target, validation loss increasing with epoch, training loss values for last 100 steps increasing etc). 

```

Comment (HarikrishnanBalagopal) [reply README.md:543]: done

File: examples/trainer_controller/README.md [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    --logging_steps 5  \
+    --logging_strategy "steps" \
+    --include_tokens_per_second  \
+    --packing False  \
+    --metric_for_best_model "loss" \
+    --load_best_model_at_end True \
+    --use_flash_attn False \
+    --trainer_controller_config_file $EXAMPLE_CONFIGS"/loss.yaml" \
+    --response_template "\n### Response:"  \
+    --dataset_text_field "output
```
Comment (ashokponkumar) [inline examples/trainer_controller/README.md:49]: We can highlight only the controller config flag

Comment (seshapad) [reply examples/trainer_controller/README.md:49]: @ashokponkumar Markdown does not allow for highlighting in code block.

File: examples/trainer_controller/README.md [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    - name: loss
+        class: Loss
+    controllers:
+    - name: loss-controller
+        triggers:
+        - on_log
+        rule: loss < 1.0
+        operations:
+        - hfcontrols.should_training_stop
+    ```
```
[the remark this answers is not in the document]
Comment (seshapad) [reply examples/trainer_controller/README.md:64]: Addressed

**Comment by seshapad** [comment]
@HarikrishnanBalagopal we dont need another directory for [trainer-controller](https://github.com/HarikrishnanBalagopal/fms-hf-tuning/tree/docs/trainercontroller/examples/trainer_controller) in the examples directory. 
Please move the readme [here](https://github.com/HarikrishnanBalagopal/fms-hf-tuning/tree/docs/trainercontroller/examples/trainercontroller_configs)

File: README.md (hunk 2 above)
Comment (seshapad) [reply README.md]: > why will anyone open details when they dont even know what is Trainer Controller? :)

Addressed. We have added a preamble to the trainer controller in the main [readme](https://github.com/foundation-model-stack/fms-hf-tuning/blob/1ca546d5050e1f6d4ba52441343fb04dc2423b89/README.md#trainer-controller-framework).

File: examples/trainer_controller/README.md (hunk 6 above)
Comment (seshapad) [reply examples/trainer_controller/README.md]: This is explained in the **Metrics** bullet point in this [section](https://github.com/foundation-model-stack/fms-hf-tuning/blob/1ca546d5050e1f6d4ba52441343fb04dc2423b89/examples/trainercontroller_configs/Readme.md#usage).

File: examples/trainer_controller/README.md (hunk 5 above)
Comment (seshapad) [reply examples/trainer_controller/README.md]: - List of metrics is provided in this [section](https://github.com/foundation-model-stack/fms-hf-tuning/blob/1ca546d5050e1f6d4ba52441343fb04dc2423b89/examples/trainercontroller_configs/Readme.md#usage).
- List of examples is also specified in the readme as a hyperlink in the above section of the readme. Its also accessible in the repo [here](https://github.com/foundation-model-stack/fms-hf-tuning/tree/1ca546d5050e1f6d4ba52441343fb04dc2423b89/examples/trainercontroller_configs).

File: examples/trainer_controller/README.md (hunk 4 above)
Comment (seshapad) [reply examples/trainer_controller/README.md]: There is only flag that needs to be added to the CLI to avail the trainer controller features. Its explained in point 1 of this [section](https://github.com/foundation-model-stack/fms-hf-tuning/tree/1ca546d5050e1f6d4ba52441343fb04dc2423b89/examples/trainercontroller_configs#usage) in the readme.

File: examples/trainer_controller/README.md (hunk 3 above)
Comment (seshapad) [reply examples/trainer_controller/README.md]: List of operations is provided in the **Operations** sub-bullet point in second point of this [section](https://github.com/foundation-model-stack/fms-hf-tuning/tree/1ca546d5050e1f6d4ba52441343fb04dc2423b89/examples/trainercontroller_configs#usage) in the readme.

File: examples/trainer_controller/README.md (hunk 4 above)
Comment (seshapad) [reply examples/trainer_controller/README.md]: Addressed. The trainer controller configuration YAML is listed in point 2 of this [section](https://github.com/foundation-model-stack/fms-hf-tuning/tree/1ca546d5050e1f6d4ba52441343fb04dc2423b89/examples/trainercontroller_configs#usage).

File: examples/trainer_controller/README.md (hunk 1 above)
Comment (seshapad) [reply examples/trainer_controller/README.md]: Answered [here](https://github.com/foundation-model-stack/fms-hf-tuning/pull/214#discussion_r1673920003) and [here](https://github.com/foundation-model-stack/fms-hf-tuning/pull/214#discussion_r1673916935).

File: README.md (hunk 2 above)
Comment (seshapad) [reply README.md]: Answered [here](https://github.com/foundation-model-stack/fms-hf-tuning/pull/214#discussion_r1673921949).

Comment (seshapad) [reply README.md]: > here please add a example of how to pass the YAML to training loop . Example: Sample YAML can look like this
> 
> ```
> controller-metrics:
>   - name: loss
>     class: Loss
> controllers:
>   - name: loss-controller
>     triggers:
>       - on_log
>     rule: "loss < 1.0"
>     operations:
>       - hfcontrols.should_training_stop
> ```
> 
> To use it with training , command with parameter needed . Pass parameter `trainer_controller_file ` with YAML file. (even i dont know what is the parameter and how to use from these docs)
> 
> ```
> `accelerate launch \
> --num_processes=4
> --config_file fixtures/accelerate_fsdp_defaults.yaml \
> tuning/sft_trainer.py  \
> --model_name_or_path $MODEL_PATH  \
> --training_data_path $TRAIN_DATA_PATH  \
> --output_dir $OUTPUT_PATH  \
> --num_train_epochs 5  \
> --per_device_train_batch_size 4  \
> --learning_rate 1e-5  \
> --response_template "\n### Label:"  \
> --dataset_text_field "output" \
> --peft_method "None"
> ---trainer_controller_file = sample_yaml
> ```
> 
> then say for details in structure of YAML and what kind of metrics you can pass , refer to [section ](examples/trainer_controller/README.md)

Answered https://github.com/foundation-model-stack/fms-hf-tuning/pull/214#discussion_r1673921949.

Comment (seshapad) [reply README.md]: > can you move few lines explanation here `A framework for controlling the trainer loop using user-defined rules and metrics.` . This can help with optimal stopping of training loop , where user could specify rules like stop training if loss below a threshold

Please check this reply https://github.com/foundation-model-stack/fms-hf-tuning/pull/214#discussion_r1673912149

**Review by kmehant — APPROVED** [review]
PR adds documentation for using trainer controller. Since all the review comments have been addressed, am merging this PR. 

For any additional feedback given, will be addressed in a separate PR. Thanks.
